File: tests/equipment.rs
//! G3/G4 — wearing things, and what that does to your numbers.
//!
//! `Mobile.equipment` was a `slot -> item` map that nothing ever wrote, and the
//! `armour` component's `defense`, `resist` and `stat_bonus` fields had no
//! reader anywhere. `combat_d.round()` ran the `damage_taken` pipeline
//! faithfully and no armour handler was ever registered in it, so **armour
//! never mitigated anything**.
//!
//! Equipping goes through the documented `equip:<slot>` source pattern rather
//! than through a second mechanism, so the assertions here are also assertions
//! about `set_source_effects` being idempotent and about `persist = false`
//! meaning what it says.

mod common;

use common::RealVm;

/// The base case: put something on, and it is on.
#[test]
fn a_player_can_wear_and_remove_armour() {
    let mut vm = RealVm::boot_real_mudlib(0);

    vm.command("spawn leather_jerkin");
    let out = vm.command("wear jerkin");
    assert!(out.contains("You wear"), "wear said nothing:\n{out}");
    assert!(out.contains("chest"), "it should name the slot:\n{out}");

    let eq = vm.command("equipment");
    assert!(eq.contains("chest") && eq.contains("jerkin"), "not shown as worn:\n{eq}");
    assert!(eq.contains("defense"), "armour should say what it is worth:\n{eq}");

    // Worn means not loose in the pack.
    assert!(
        !vm.command("inventory").contains("jerkin"),
        "a worn item should not still be listed as carried"
    );

    let out = vm.command("remove chest");
    assert!(out.contains("You stop using"), "remove said nothing:\n{out}");
    assert!(vm.command("inventory").contains("jerkin"), "it should come back to the pack");
}

/// The refusal path — one rule in `lib/requires.lua`, shared by weapons and
/// armour, which is why there is one message rather than two.
#[test]
fn a_requirement_refuses_with_a_reason() {
    let mut vm = RealVm::boot_real_mudlib(0);

    vm.command("spawn iron_greatsword");
    let out = vm.command("wield greatsword");
    assert!(
        out.contains("16 strength"),
        "the refusal should name what is missing:\n{out}"
    );
    assert!(
        !vm.command("equipment").contains("greatsword"),
        "a refused item must not end up equipped anyway"
    );

    // Meet it and the same command works. `affect learn` writes a trait base,
    // which is exactly what a strength buff would end up changing.
    vm.command("affect learn strength 18");
    let out = vm.command("wield greatsword");
    assert!(out.contains("You wield"), "still refused at 18 strength:\n{out}");
}

/// A two-handed weapon takes both hands, in either order.
#[test]
fn a_two_handed_weapon_clears_the_offhand() {
    let mut vm = RealVm::boot_real_mudlib(0);
    vm.command("affect learn strength 18");

    vm.command("spawn oak_buckler");
    vm.command("spawn iron_greatsword");

    vm.command("wear buckler");
    assert!(vm.command("equipment").contains("buckler"));

    let out = vm.command("wield greatsword");
    assert!(out.contains("You wield"), "{out}");
    assert!(
        out.contains("stop using") && out.contains("buckler"),
        "displacing the shield should be said out loud, not done silently:\n{out}"
    );

    let eq = vm.command("equipment");
    assert!(eq.contains("greatsword"));
    assert!(!eq.contains("buckler"), "the offhand should be empty:\n{eq}");

    // And the other way round: putting a shield back takes the sword out.
    let out = vm.command("wear buckler");
    assert!(out.contains("You wear"), "{out}");
    let eq = vm.command("equipment");
    assert!(eq.contains("buckler"));
    assert!(
        !eq.contains("greatsword"),
        "both hands were on the greatsword; a shield cannot join it:\n{eq}"
    );
}

/// `stat_bonus` becomes a real `equip:` effect, visible in `score`, and gone
/// when the item comes off. This is the field that had no reader at all.
#[test]
fn a_stat_bonus_is_an_equip_effect_that_score_shows() {
    let mut vm = RealVm::boot_real_mudlib(0);

    let before = vm.command("affect traits");
    let intelligence_line = |out: &str| -> String {
        out.lines()
            .find(|l| l.contains("intelligence"))
            .unwrap_or_else(|| panic!("no intelligence row in:\n{out}"))
            .to_string()
    };
    assert!(
        intelligence_line(&before).contains("10"),
        "expected a base of 10: {}",
        intelligence_line(&before)
    );

    vm.command("spawn scholar_circlet");
    vm.command("wear circlet");

    let after = intelligence_line(&vm.command("affect traits"));
    assert!(
        after.contains("10") && after.contains("12"),
        "expected base 10 and effective 12 while the circlet is worn, got: {after}"
    );

    // The effect names its source, so `effects` explains where the number came
    // from rather than leaving a mystery +2.
    let effects = vm.command("affect list");
    assert!(
        effects.contains("equip:head"),
        "the aura should be sourced to the slot:\n{effects}"
    );

    vm.command("remove head");
    let after = intelligence_line(&vm.command("affect traits"));
    assert!(
        !after.contains("12"),
        "the bonus outlived the item that granted it: {after}"
    );
}

/// **G4** — armour finally mitigates, in the `reduce` phase, and a resist table
/// is applied by damage type.
#[test]
fn armour_reduces_damage_and_resist_is_by_type() {
    let mut vm = RealVm::boot_real_mudlib(0);

    // Unarmoured baseline, through the real pipeline.
    let plain = vm.command("affect damage 30");
    assert!(plain.contains("30 requested, 30 dealt"), "unmitigated:\n{plain}");

    vm.command("affect heal 500");
    vm.command("spawn leather_jerkin");
    vm.command("wear jerkin");

    // defense 3, in the reduce phase.
    let armoured = vm.command("affect damage 30");
    assert!(
        armoured.contains("30 requested, 27 dealt"),
        "leather (defense 3) should take 3 off a 30-point hit:\n{armoured}"
    );

    // A resist table meets a damage type. The warded cloak blunts magic by 6
    // and does nothing at all to a sword — which is the whole reason
    // `damage_type` exists.
    vm.command("affect heal 500");
    vm.command("spawn warded_cloak");
    vm.command("wear cloak");

    let physical = vm.command("affect damage 30 physical");
    assert!(
        physical.contains("30 requested, 26 dealt"),
        "jerkin 3 + cloak 1 = 4 against physical:\n{physical}"
    );

    vm.command("affect heal 500");
    let magical = vm.command("affect damage 30 magic");
    assert!(
        magical.contains("30 requested, 20 dealt"),
        "jerkin 3 + cloak 1 + cloak's magic resist 6 = 10 against magic:\n{magical}"
    );
}

/// Two pieces of armour are two independent `equip:` sources, so taking one off
/// leaves the other's mitigation alone.
#[test]
fn each_slot_is_its_own_source() {
    let mut vm = RealVm::boot_real_mudlib(0);

    vm.command("spawn leather_jerkin");
    vm.command("spawn warded_cloak");
    vm.command("wear jerkin");
    vm.command("wear cloak");

    vm.command("affect heal 500");
    assert!(vm.command("affect damage 30 physical").contains("26 dealt"));

    vm.command("remove back");
    vm.command("affect heal 500");
    assert!(
        vm.command("affect damage 30 physical").contains("27 dealt"),
        "removing the cloak should take away exactly the cloak's contribution"
    );
}

/// `equip:` effects are `persist = false` and are rebuilt from what is worn on
/// login. Persisting the aura as well would be a second copy of the truth that
/// can disagree with the first.
#[test]
fn the_equip_aura_is_derived_and_never_written() {
    let mut vm = RealVm::boot_real_mudlib_with_probe();

    vm.eval("_E = require('lib.equipment')").unwrap();
    vm.eval(
        "_p = { char_id = 71, name = 'Wearer', inventory = {}, equipment = {}, \
                stats = { strength = 10, dexterity = 10, constitution = 10, \
                          intelligence = 10, wisdom = 10, level = 1, hp = 100, mp = 50 }, \
                send = function() end }",
    )
    .unwrap();
    vm.eval("setmetatable(_p, { __index = require('lib.mobile') }) return 'ok'").unwrap();
    vm.eval("DAEMON.trait.attach(_p)").unwrap();

    vm.eval("_circlet = DAEMON.items.spawn('scholar_circlet', nil)").unwrap();
    vm.eval("_E.equip(_p, _circlet, DAEMON.items.resolve(_circlet))").unwrap();
    assert_eq!(
        vm.eval("return _p:trait('intelligence')").unwrap(),
        "12",
        "the aura did not apply"
    );

    // Idempotent: refreshing twice must not double it. This is the property
    // that lets login call it without working out what it did last time.
    vm.eval("_E.refresh_all(_p) _E.refresh_all(_p) return 'refreshed'").unwrap();
    assert_eq!(
        vm.eval("return _p:trait('intelligence')").unwrap(),
        "12",
        "refreshing stacked the aura on top of itself"
    );

    // Nothing about it is written to the persisted namespace.
    assert_eq!(
        vm.eval(
            "local n = 0 for _, e in ipairs(DAEMON.effect.active(_p)) do \
             if e.ns == 'effects' then n = n + 1 end end return n"
        )
        .unwrap(),
        "0",
        "an equip aura reached the persisted effects namespace"
    );

    // And swapping to a *different* item in the same slot replaces the amount
    // rather than keeping the old one, which is what `set_source_effects`
    // matching by definition id would do if it were called only once.
    vm.eval(
        "DAEMON.items.register(require('lib.armor'){ id = 'test_crown', \
           short = 'a crown', slot = 'head', stat_bonus = { intelligence = 5 } })",
    )
    .unwrap();
    vm.eval("_crown = DAEMON.items.spawn('test_crown', nil)").unwrap();
    vm.eval("_E.equip(_p, _crown, DAEMON.items.resolve(_crown))").unwrap();
    assert_eq!(
        vm.eval("return _p:trait('intelligence')").unwrap(),
        "15",
        "swapping head gear kept the old item's bonus"
    );

    vm.eval("_E.unequip(_p, 'head') return 'off'").unwrap();
    assert_eq!(
        vm.eval("return _p:trait('intelligence')").unwrap(),
        "10",
        "the bonus outlived the item"
    );
}

/// A `stat_bonus` aimed at a gauge is refused with a message naming the item's
/// field, rather than producing a warning about an effect definition nobody
/// wrote. Effects modify attributes and derived traits, never gauges.
#[test]
fn a_stat_bonus_cannot_target_a_gauge() {
    let mut vm = RealVm::boot_real_mudlib_with_probe();

    vm.eval("_E = require('lib.equipment')").unwrap();
    vm.eval(
        "DAEMON.items.register(require('lib.armor'){ id = 'bad_amulet', \
           short = 'an amulet', slot = 'neck', stat_bonus = { hp = 50 } })",
    )
    .unwrap();
    vm.eval(
        "_p = { char_id = 72, inventory = {}, equipment = {}, \
                stats = { constitution = 10, level = 1, hp = 100 }, \
                send = function() end }",
    )
    .unwrap();
    vm.eval("setmetatable(_p, { __index = require('lib.mobile') }) return 'ok'").unwrap();

    vm.eval("_a = DAEMON.items.spawn('bad_amulet', nil)").unwrap();
    // Equipping still succeeds — the item is wearable, one of its fields is
    // just meaningless. Refusing the whole item would make one bad number
    // unwearable rather than inert.
    assert_eq!(
        vm.eval("return tostring(_E.equip(_p, _a, DAEMON.items.resolve(_a)))").unwrap(),
        "true"
    );
    assert_eq!(
        vm.eval("return tostring(DAEMON.effect.get_def('equip_trait_hp'))").unwrap(),
        "nil",
        "an effect targeting a gauge should never have been defined"
    );
}

File: mudlib/lib/equipment.lua
-- mudlib/lib/equipment.lua — Wearing and wielding, and what that does to your
-- numbers.
--
-- `Mobile.equipment` was a `slot -> item` map that **nothing ever wrote**, and
-- the `armour` component's `defense`, `resist` and `stat_bonus` fields had no
-- reader anywhere. Combat ran the `damage_taken` pipeline faithfully and no
-- armour handler was ever registered in it, so armour never mitigated anything.
--
-- Equipping goes through the documented `equip:` source pattern rather than
-- through a second mechanism:
--
--     DAEMON.effect.set_source_effects(entity, "equip:chest", specs)
--
-- *The effects from this source are now exactly these.* Idempotent, so it is
-- safe on every login and every slot change without working out what it did
-- last time, and `persist = false` means nothing is ever written — the aura is
-- rebuilt from what is worn, which is the only copy that can be wrong.
--
-- ─── Why the effect definitions are generated ────────────────────────────────
--
-- An effect's hooks are fixed when it is *defined*, and a trait modifier is a
-- `trait:<id>` hook. A single "equipment aura" definition therefore cannot
-- modify strength on one character and wisdom on another — the hook name is
-- part of the definition, not of the instance.
--
-- So one definition is generated per trait any gear actually touches,
-- `equip_trait_<id>`, with the amount carried per instance in `state`. The
-- population is bounded by how many distinct traits the game's gear modifies,
-- which is small, and `effect_d.define` already handles being called at runtime
-- (it bumps the generation and invalidates every memo).
--
-- Protection is different and needs only one definition: `damage_taken` is a
-- single hook whatever the damage type, so `equip_protection` reads the worn
-- piece's defence and resist table out of its own instance state.

local Armor      = require('components.armor')
local Weapon     = require('components.weapon')
local Container  = require('components.container')
local Requires   = require('components.requires')
local Components = require('components')

local M = {}

--- The slots this mudlib knows about, in the order `equipment` lists them.
--- A game may use fewer; using more means adding one here, which is the point
--- of it being a mudlib list rather than a game one.
M.SLOTS = {
    "head", "neck", "chest", "back", "hands", "waist",
    "legs", "feet", "weapon", "offhand", "light", "ring",
}

local SLOT_SET = {}
for _, slot in ipairs(M.SLOTS) do SLOT_SET[slot] = true end

--- Slots a two-handed weapon occupies. Wielding one has to clear the offhand,
--- or a shield keeps working while both hands are on a greatsword.
M.TWO_HANDED_SLOTS = { "weapon", "offhand" }

local function log_error(message)
    log("error", message)
    if DAEMON and DAEMON.journal then pcall(DAEMON.journal.error, message) end
end

--- @param slot any
--- @return boolean
function M.is_slot(slot)
    return type(slot) == "string" and SLOT_SET[slot] == true
end

-- ─── Generated effect definitions ────────────────────────────────────────────

--- The definition id for the aura that modifies one trait.
local function trait_def_id(trait_id)
    return "equip_trait_" .. trait_id
end

--- Make sure an effect exists that adds `state.amount` to one trait.
---
--- Defined on demand rather than up front, because the mudlib does not know
--- which traits a game's gear will touch and enumerating every defined trait
--- would create hundreds of effect definitions nobody uses.
--- @param trait_id string
--- @return string|nil  the definition id, or nil if it could not be defined
local function ensure_trait_effect(trait_id)
    if not (DAEMON and DAEMON.effect) then return nil end
    local id = trait_def_id(trait_id)
    if DAEMON.effect.get_def(id) then return id end

    -- A gauge or a counter cannot be modified — `effect_d.define` refuses a
    -- `trait:hp` hook by name, at registration time. Checking here turns that
    -- refusal into a message naming the item's field rather than a warning
    -- about a definition the author never wrote.
    if DAEMON.trait and DAEMON.trait.get_def then
        local def = DAEMON.trait.get_def(trait_id)
        if def and (def.kind == "gauge" or def.kind == "counter") then
            log_error("EQUIPMENT: a stat_bonus cannot target '" .. trait_id
                .. "', which is a " .. def.kind
                .. " — raise the trait that is its max instead")
            return nil
        end
    end

    local ok = DAEMON.effect.define({
        id      = id,
        label   = "Equipment",
        -- Never written. The aura is derived from what is worn, and what is
        -- worn is saved; persisting the aura as well would be a second copy
        -- that can disagree with the first.
        persist = false,
        hooks   = {
            [trait_id] = {
                hook  = "trait:" .. trait_id,
                phase = "add",
                fn    = function(ev, ctx)
                    local amount = ctx.inst.state and ctx.inst.state.amount
                    if type(amount) == "number" then
                        ev.amount = (ev.amount or 0) + amount
                    end
                end,
            },
        },
    })
    return ok and id or nil
end

--- The one definition that turns worn armour into real mitigation (G4).
---
--- `reduce` phase, so a percentage multiplier applied by something else lands
--- first — which is the ordering `effects.md` argues for and the reason a
--- 30-point hit against stoneskin yields the documented 20 and not 21.
local function ensure_protection_effect()
    if not (DAEMON and DAEMON.effect) then return nil end
    if DAEMON.effect.get_def("equip_protection") then return "equip_protection" end

    local ok = DAEMON.effect.define({
        id      = "equip_protection",
        label   = "Armour",
        persist = false,
        hooks   = {
            damage_taken = {
                hook  = "damage_taken",
                phase = "reduce",
                fn    = function(ev, ctx)
                    local state = ctx.inst.state
                    if type(state) ~= "table" then return end

                    local reduction = tonumber(state.defense) or 0
                    -- Resist is looked up by the damage type on the event, so
                    -- a warded cloak blunts the wisp and does nothing at all
                    -- against a sword. A negative entry is a weakness and
                    -- increases the number, which is the same arithmetic.
                    local by_type = type(state.resist) == "table"
                        and tonumber(state.resist[ev.damage_type or "physical"]) or nil
                    if by_type then reduction = reduction + by_type end

                    ev.amount = (ev.amount or 0) - reduction
                end,
            },
        },
    })
    return ok and "equip_protection" or nil
end

-- ─── What one worn piece contributes ─────────────────────────────────────────

--- The effect specs a single equipped item produces.
--- @param item table  the resolved item
--- @return table  array of `set_source_effects` specs
local function specs_for(item)
    local specs = {}
    if type(item) ~= "table" then return specs end

    -- Whatever the item's components contribute. Nothing here names a
    -- component: armour's mitigation lives in `components/armor.lua` where the
    -- rest of armour does, and a component added later is picked up by
    -- existing. The two effect definitions stay here because they are
    -- equipment's to create, so they are handed over rather than reached for.
    for _, spec in ipairs(Components.equip_specs(item, {
        trait_effect      = ensure_trait_effect,
        protection_effect = ensure_protection_effect,
    })) do
        specs[#specs + 1] = spec
    end

    -- A weapon may carry stat bonuses too — a sword of strength is an ordinary
    -- thing and needs no separate mechanism.
    if type(item.stat_bonus) == "table" then
        for trait_id, amount in pairs(item.stat_bonus) do
            if type(amount) == "number" and amount ~= 0 then
                local def_id = ensure_trait_effect(trait_id)
                if def_id then
                    specs[#specs + 1] = { def = def_id, state = { amount = amount } }
                end
            end
        end
    end

    return specs
end

--- Rebuild one slot's aura from what is in it.
---
--- Cleared first and then set, in two calls. `set_source_effects` matches by
--- definition id, so swapping one chest piece for another that modifies the
--- same trait would otherwise see "an instance of `equip_trait_strength` from
--- `equip:chest` already exists" and keep the *old* item's amount.
--- @param entity table
--- @param slot string
--- @param item table|nil  nil clears the slot's aura
function M.refresh_slot(entity, slot, item)
    if not (DAEMON and DAEMON.effect) then return end
    local source = "equip:" .. slot

    local ok, err = pcall(DAEMON.effect.set_source_effects, entity, source, {})
    if not ok then
        log_error("EQUIPMENT: clearing '" .. source .. "' failed: " .. tostring(err))
        return
    end
    if not item then return end

    local specs = specs_for(item)
    if #specs == 0 then return end

    local sok, serr = pcall(DAEMON.effect.set_source_effects, entity, source, specs)
    if not sok then
        log_error("EQUIPMENT: applying '" .. source .. "' failed: " .. tostring(serr))
    end
end

--- Rebuild every slot's aura from scratch.
---
--- Called on login, because `equip:` effects are `persist = false` and so are
--- gone by design — what is worn is saved, and the aura is derived from it.
--- @param entity table
function M.refresh_all(entity)
    if type(entity) ~= "table" or type(entity.equipment) ~= "table" then return end
    for _, slot in ipairs(M.SLOTS) do
        local entry = entity.equipment[slot]
        local item = entry and DAEMON and DAEMON.items and DAEMON.items.resolve(entry)
        M.refresh_slot(entity, slot, item)
    end
end

-- ─── The two verbs, as one operation each ────────────────────────────────────

--- Which slot an item goes in, and what else it displaces.
--- @param item table
--- @return string|nil slot, table|nil also_occupies
function M.slot_for(item)
    if type(item) ~= "table" then return nil end
    local slot = item.slot
    if not M.is_slot(slot) then return nil end
    if slot == "weapon" and Weapon.is(item) and item.weapon.two_handed then
        return slot, { "offhand" }
    end
    return slot, nil
end

--- Put an item on.
---
--- Returns rather than sends: the caller knows whether it is `wear` or `wield`
--- and what the room should be told, and a library that wrote to a socket would
--- be untestable without one.
--- @param entity table
--- @param entry table   the instance
--- @param item table    the resolved item
--- @return boolean ok, string|nil why, table|nil displaced  array of entries removed
function M.equip(entity, entry, item)
    local slot, also = M.slot_for(item)
    if not slot then
        return false, "You cannot wear or wield that."
    end
    if not item.equippable then
        return false, "You cannot wear or wield that."
    end

    -- The one refusal path, for level, strength and dexterity alike. Read
    -- through the entity rather than its stored stats, so a strength buff
    -- genuinely lets you lift the greatsword.
    local met, why = Requires.met(item, entity)
    if not met then return false, why end

    entity.equipment = entity.equipment or {}

    -- Anything already in the way comes off first, including the offhand when
    -- a two-handed weapon goes on.
    local displaced = {}
    local to_clear = { slot }
    for _, extra in ipairs(also or {}) do to_clear[#to_clear + 1] = extra end
    -- A shield in the offhand also blocks a two-handed weapon in reverse: if
    -- the weapon slot holds one, filling the offhand has to clear it.
    if slot == "offhand" then
        local held = entity.equipment.weapon
        local held_item = held and DAEMON and DAEMON.items and DAEMON.items.resolve(held)
        if held_item and Weapon.is(held_item) and held_item.weapon.two_handed then
            to_clear[#to_clear + 1] = "weapon"
        end
    end

    for _, s in ipairs(to_clear) do
        local occupant = entity.equipment[s]
        if occupant then
            local removed, rwhy = M.unequip(entity, s)
            if not removed then return false, rwhy end
            displaced[#displaced + 1] = occupant
        end
    end

    -- Out of the inventory array. An item cannot be both worn and loose in the
    -- pack: leaving it in both meant `inventory` listed what you were wearing,
    -- and — worse — that `drop sword` could put a wielded sword on the floor
    -- while `equipment` went on reporting it in your hand.
    if type(entity.inventory) == "table" then
        for i, e in ipairs(entity.inventory) do
            if e == entry then
                table.remove(entity.inventory, i)
                break
            end
        end
    end

    entity.equipment[slot] = entry
    M.refresh_slot(entity, slot, item)

    local _, hook_ran = pcall(function()
        return require('lib.carry').fire_hook(item, "on_equip", entity.char_id)
    end)
    local _ = hook_ran

    if DAEMON and DAEMON.event then
        pcall(DAEMON.event.emit, "item.equipped", {
            char_id = entity.char_id, slot = slot,
            instance_id = type(entry) == "table" and entry.id,
            template_id = type(entry) == "table" and entry.template,
        })
    end

    return true, nil, displaced
end

--- Take an item off and put it back in the inventory.
--- @param entity table
--- @param slot string
--- @return boolean ok, string|nil why, table|nil entry
function M.unequip(entity, slot)
    if type(entity.equipment) ~= "table" then return false, "You are not wearing that." end
    local entry = entity.equipment[slot]
    if not entry then return false, "You are not wearing anything there." end

    local item = DAEMON and DAEMON.items and DAEMON.items.resolve(entry)

    -- A cursed item would refuse here. Nothing does yet; the hook is where it
    -- would go, and it goes before the state changes so a refusal leaves
    -- nothing half-done.
    entity.equipment[slot] = nil
    M.refresh_slot(entity, slot, nil)

    entity.inventory = entity.inventory or {}
    entity.inventory[#entity.inventory + 1] = entry

    if item then
        require('lib.carry').fire_hook(item, "on_remove", entity.char_id)
    end
    if DAEMON and DAEMON.event then
        pcall(DAEMON.event.emit, "item.unequipped", {
            char_id = entity.char_id, slot = slot,
            instance_id = type(entry) == "table" and entry.id,
            template_id = type(entry) == "table" and entry.template,
        })
    end

    return true, nil, entry
end

--- The equipped item in one slot, resolved.
--- @return table|nil entry, table|nil item
function M.worn(entity, slot)
    local entry = type(entity.equipment) == "table" and entity.equipment[slot]
    if not entry then return nil, nil end
    return entry, DAEMON and DAEMON.items and DAEMON.items.resolve(entry)
end

--- Every occupied slot, in `M.SLOTS` order.
--- @return table  array of { slot, entry, item }
function M.all_worn(entity)
    local out = {}
    for _, slot in ipairs(M.SLOTS) do
        local entry, item = M.worn(entity, slot)
        if entry then out[#out + 1] = { slot = slot, entry = entry, item = item } end
    end
    return out
end

--- Total encumbrance from worn armour, for whatever a game wants to do with it.
--- @return number
function M.encumbrance(entity)
    local total = 0
    for _, worn in ipairs(M.all_worn(entity)) do
        total = total + (Armor.encumbrance(worn.item) or 0)
    end
    return total
end

-- Re-exported so a command needing "is this a container" has one import.
M.Container = Container

return M
